### PR TITLE
Np 9226 lti integration with redirect

### PR DESCRIPTION
**Pull request by truhacevkir**

### CODE REVIEW

File: docs/openapi.yaml [hunk 1]
```diff
@@ -5,59 +5,66 @@ info:
...
                 $ref: '#/components/schemas/Error'
-        '400':
-          description: Bad request
+        '401':
+          description: Bad lti launch request
           content:
-            application/json:
+            text/plain:
               schema:
                 $ref: '#/components/schemas/Error'
```
Comment (ketilaa) [inline docs/openapi.yaml:57]: Her bør type bare være string, eller kanskje bedre å endre content til application/json og faktisk mappe body fra LtiLaunchRequest til en ApiGatawayException hvis status er 401?

File: docs/openapi.yaml [hunk 2]
```diff
@@ -5,59 +5,66 @@ info:
...
-            application/json:
+            application/xml:
+              schema:
+                $ref: '#/components/schemas/Cartridge'
+            text/html:
               schema:
-                $ref: '#/components/schemas/Response'
+                $ref: '#/components/schemas/HTML'
+        '304':
+          description: ''
```
Comment (ketilaa) [inline docs/openapi.yaml]: Litt faktisk beskrivelse kan være greit her.

File: docs/openapi.yaml [hunk 3]
```diff
@@ -5,59 +5,66 @@ info:
...
         uri:
-          Fn::Sub: arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${DlrLmsServiceFunction.Arn}/invocations
-        httpMethod: POST
+          Fn::Sub: arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${LmsCanvasLaunchFunction.Arn}/invocations
+        httpMethod: GET
         type: AWS_PROXY
       responses:
-        '201':
+        '200':
           description: ''
```
Comment (ketilaa) [inline docs/openapi.yaml]: Say something in description about when to expect the different content types.

File: docs/openapi.yaml [hunk 4]
```diff
@@ -5,59 +5,66 @@ info:
   license:
     name: MIT
 paths:
-  /lms:
-    post:
+  /lms/lti/canvas/{serviceId}:
+    parameters:
+      - in: path
+        name: serviceId
```
Comment (ketilaa) [inline docs/openapi.yaml:11]: Say something about valid values. Maybe even create a string enum for the valid values.

File: lms-canvas/src/main/java/no/sikt/nva/lms/CanvasLaunchHandler.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+import nva.commons.apigateway.ApiGatewayHandler;
+import nva.commons.apigateway.RequestInfo;
+import sikt.lti.tp.aws.ApiGatewayLambdaLaunchHandler;
+
+public class CanvasLaunchHandler extends ApiGatewayHandler<Void, String> {
+
+    private static final URI DLR_BASE_URL = URI.create("https://dlr.unit.no");
```
Comment (ketilaa) [inline lms-canvas/src/main/java/no/sikt/nva/lms/CanvasLaunchHandler.java]: DLR_BASE_URL and API_HOST will differ in different environments. Should be read from environment.

File: lti-tool-provider/src/main/java/sikt/lti/tp/LtiLaunchResult.java [hunk 6]
```diff
@@ -1,6 +1,9 @@
 package sikt.lti.tp;
 
-public class LtiLaunchResult {
+import no.unit.nva.commons.json.JsonSerializable;
+
+public class LtiLaunchResult implements JsonSerializable {
```
Comment (ketilaa) [inline lti-tool-provider/src/main/java/sikt/lti/tp/LtiLaunchResult.java]: Not needed to implement JsonSerializable?

File: lti-tool-provider/src/main/java/sikt/lti/tp/aws/ApiGatewayLambdaLaunchHandler.java [hunk 7]
```diff
@@ -1,16 +1,16 @@
 package sikt.lti.tp.aws;
 
-import com.amazonaws.services.lambda.runtime.events.APIGatewayV2HTTPEvent;
 import java.net.URI;
 import java.util.Set;
+import nva.commons.apigateway.RequestInfo;
```
Comment (ketilaa) [inline lti-tool-provider/src/main/java/sikt/lti/tp/aws/ApiGatewayLambdaLaunchHandler.java:5]: I guess you can remove dependency to was-events in gradle with this change?

File: template.yaml [hunk 8]
```diff
@@ -50,44 +50,44 @@ Resources:
...
+  LmsCanvasGatewayAPI:
     Type: AWS::Serverless::Api
     Properties:
       StageName: v1
       EndpointConfiguration:
         Type: REGIONAL
       MethodSettings:
         - ResourcePath: /lms
-          HttpMethod: POST
+          HttpMethod: GET
```
Comment (ketilaa) [inline template.yaml]: I think it might have to be a post here? I think api gateway always POST to the lambda function.

**Review by ketilaa — CHANGES_REQUESTED** [review]

File: docs/openapi.yaml (hunk 1 above)
Comment (truhacevkir) [reply docs/openapi.yaml:57]: har fikset

File: docs/openapi.yaml (hunk 2 above)
Comment (truhacevkir) [reply docs/openapi.yaml]: lagt til

File: docs/openapi.yaml (hunk 3 above)
Comment (truhacevkir) [reply docs/openapi.yaml]: endret

File: docs/openapi.yaml (hunk 4 above)
Comment (truhacevkir) [reply docs/openapi.yaml:11]: added enums, with nullable: true

File: lti-tool-provider/src/main/java/sikt/lti/tp/LtiLaunchResult.java (hunk 6 above)
Comment (truhacevkir) [reply lti-tool-provider/src/main/java/sikt/lti/tp/LtiLaunchResult.java]: no need yes, my fault

File: lti-tool-provider/src/main/java/sikt/lti/tp/aws/ApiGatewayLambdaLaunchHandler.java (hunk 7 above)
Comment (truhacevkir) [reply lti-tool-provider/src/main/java/sikt/lti/tp/aws/ApiGatewayLambdaLaunchHandler.java:5]: removed, not in use longer

File: template.yaml (hunk 8 above)
Comment (truhacevkir) [reply template.yaml]: have changed

**Comment by truhacevkir** [comment]
Secret key/value/location has to be changed for prod

**Pull request merged**